### PR TITLE
Async volume creation

### LINKED ISSUE #1368: BaggageClaim volume creation should have an async API
We can bump the timeout as much as we want but given that it can be O(N) based on the strategy no timeout will ever suffice for everything.

Let's make it async so we don't even need a crazy timeout.

### PR DESCRIPTION
**Pull request by SHyx0rmZ**
This aims to close concourse/concourse#1368. It introduces three new API routes: `POST /volumes-async`, `GET /volumes-async/:handle` and `DELETE /volumes-async/:handle`.

Upon `POST`ing to `/volumes-async`, the `api.VolumeServer` will add a `volume.Promise` to it's list of promises and then return the handle of the promise to the client.

The client can then `GET` `/volumes-async/:handle`, which will either return a `204` with no body, or a `200` containing the traditional `volume.Volume` encoded as JSON.

Sending a request with method `DELETE` to `/volumes-async/:handle` will delete a promise from the volume server if it is no longer needed, or mark the promise as canceled, in which case the volume server will continue to call `DeleteVolume`, should the `CreateVolume` succeed.

### CODE REVIEW

**Comment by jtarchie** [comment]
@SHyx0rmZ, thanks for the PR. I've been spending the afternoon investigating (poking) it.

1. Ran it locally with a bosh virtualbox deployed concourse. It works with some example pipelines [1][2]. 
2. Ran [testflight] against the deployment. The test that exercises `timeout` on a task failed. Its total runtime was *1766.798* seconds. For reference, it takes *734.361* seconds without your changes.
3. Reran the test with `timeout` several times and 4/5 it failed. The one time it passed it took *27* seconds. The same test without your changes passes consistently at *~11* seconds.

I don't have any reference if it is more stable, but things are definitely taking longer. Going to dive into it more.

[1] https://github.com/concourse/testflight/blob/master/pipelines/fixtures/do.yml
[2] https://github.com/concourse/testflight/blob/master/pipelines/fixtures/simple.yml

**Comment by vito** [comment]
Awesome, thanks for doing this!

A couple things:

1. I think we can hide all this under the existing client-side API. We don't have a use case for exposing async creation up to the client, I don't think. Can we have `CreateVolume` on the client just do the async create + wait under the hood?
1. I'm guessing what @jtarchie found with the tests slowing down could be due to there always being the initial polling delay added to every volume create. Could we do an exponential backoff up to 10 seconds instead?

File: client/volume_future.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	for {
+		response, err := f.client.httpClient(f.logger).Do(request)
+		if err != nil {
+			return nil, err
+		}
+
+		if response.StatusCode == http.StatusNoContent {
+			response.Body.Close()
+
+			time.Sleep(10 * time.Second)
```
Comment (SHyx0rmZ) [inline client/volume_future.go]: Do exponential back-off here.

**Comment by SHyx0rmZ** [comment]
Personally, I feel like I wouldn't want to force this onto all existing clients, which is why I exposed new routes. This partly stems from the fact that I'm not sure that this is the right thing to do. I can't clearly see how this will resolve the timeout problem, yet.

On a (side) note: The community recently came up with the idea of adding a possibility to narrow down the set of artifacts (or paths within them) that will have to be copied to reduce the time until the response. Maybe that idea obsoletes this pull request?

**Comment by SHyx0rmZ** [comment]
@vito I updated the pull request with what I understood were the changes you suggested. I feel like I might have misunderstood exactly what you envisioned when you suggested to hide everything under the existing client-side API.

The functionality is still done inside the client, but it's no longer exposed in the Client interface. I figured we would still have to do it in the client, because this is about the HTTP requests after all.

If we want to expose the asynchronous version through the `POST /volumes` handler and get rid of the synchronous version I'd also suggest moving the routes for the future to `/promise` or `/future`.

If there still is anything I can do or you had some smart idea about all of this, let me know!

File: client/volume_future.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+		response, err := f.client.httpClient(f.logger).Do(request)
+		if err != nil {
+			return nil, err
+		}
+
+		if response.StatusCode == http.StatusNoContent {
+			response.Body.Close()
+
+			if backoffFactor != 0 {
+				time.Sleep(time.Duration(rand.Intn((1<<backoffFactor)-1)) * time.Second)
```
Comment (vito) [inline client/volume_future.go]: Seconds may be a bit too long. How about a baseline of 10ms, just to make something up? I figure at least in the common case of creating an empty volume it should complete really quickly.

I'd recommend taking a look at https://github.com/cenkalti/backoff - we use it elsewhere and is easier to use at a high level. See https://godoc.org/github.com/cenkalti/backoff#ExponentialBackOff

**Review by vito — CHANGES_REQUESTED** [review]
@SHyx0rmZ Sorry for my inattentiveness. Yeah, I just meant to keep the Go interface the same, and hide all the HTTP async stuff behind the existing interfaces. The latest changes look great! I left one comment around the backoff.

**Comment by SHyx0rmZ** [comment]
Haha, don't worry. I also took my sweet time 😆

**Comment by vito** [comment]
Tried out TestFlight and it takes the same amount of time as before. Thanks!

**Comment by SHyx0rmZ** [comment]
Happy to hear! Thanks for all the feedback!

**Pull request merged**